src/plonk.rs

```diff
@@ -198,16 +219,18 @@ impl<C: HaloCurve> Circuit<C> {
         let all_opened_values_bf: Vec<_> = all_opened_values_sf
             .into_iter()
             .map(|f| {
+                // TODO: Fix this, this can regularly fail if a public input is not in range for example.
                 C::try_convert_s2b(f)
                     .expect("For now, we assume that all opened values fit in both fields")
             })
             .collect();
 
         // Generate random v, u, and x from the transcript.
         challenger.observe_elements(&all_opened_values_bf);
-        let (v_bf, u_bf, _x) = challenger.get_3_challenges();
+        let (v_bf, u_bf) = challenger.get_2_challenges();
         let v_sf = v_bf.try_convert::<C::ScalarField>()?;
         let u_sf = u_bf.try_convert::<C::ScalarField>()?;
+        let u_curve = challenger.get_affine_point::<C>();
```

Comment:
Can this be a random constant, maybe using hash-to-point (with a constant input, or `degree` plus a constant)? That was my understanding from the paper.

Reply:
So, I think Bulletproof has the verifier sending a random scalar challenge, while Halo has the verifier sending `U` a random curve point. I didn't know if there were security issues with the prover knowing the discrete log of `U`, so I went with the Halo way. If you think it's safe I can change it back.

Reply:
Hm my understanding is that the Halo paper assumes that Bulletproofs' "Protocol 1" reduction already occurred, so I think the `U` used in Halo's Schnorr protocol should be a random constant point scaled by a challenge.

Reply:
I don't think Halo assumes Protocol 1 of Bulletproof. I think the first paragraph of Section 3.1 in the Halo paper is analogous to Protocol 1:
"The protocol takes the polynomial commitment `P`, point `x`, and claimed evaluation `v` as common inputs. In the first move the verifier sends a **random group element `U ∈ G`**. Both parties compute `P' = P + [v]U`".
Now the question is whether it matters if the prover knows the discrete log of `U`. Probably not, but to be sure I used the version where the prover doesn't know it.

Reply:
Oh I missed that bit. Will think about that some more.